sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2158,4 +2158,21 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     val mean = result.select("DecimalCol").where($"summary" === "mean")
     assert(mean.collect().toSet === Set(Row("0.0345678900000000000000000000000000000")))
   }
+
+  testQuietly("SPARK-21413: Multiple projections with CASE WHEN fails") {
+    val schema = StructType(StructField("a", IntegerType) :: Nil)
+    val df = spark.createDataFrame(sparkContext.parallelize(Seq(Row(1))), schema)
+    val df1 =
+      df.withColumn("a", when($"a" === 0, null).otherwise($"a"))
```

Comment:
maybe use a loop here?

Reply:
Sure, the following code can throw an exception
```
...
    var df1 = df
    for (i <- 1 to 10) {
      df1 = df1.withColumn("a", when($"a" === 0, null).otherwise($"a"))
    }
...
```